### ISSUE TITLE
Bug: [no-unnecessary-type-arguments] false positives in conjunctions with `any`

### ISSUE THREAD

**Issue by StyleShit** [typescript-eslint/typescript-eslint#12183]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=GYVwdgxgLglg9mABCAzgUwMpQIZTQdRigAsMYATNAUWGDWgB4AVAPgAoYwiZsAbLXGgBciJgBpEKCtVr0oAfhFswaAO4A1PiGGiAlIgC8LRADc4FfQG8AUIkS80UU1rSHEnbnwF4A3NduIEAgoTuhQmrzabspqEdoiTPpGiDZ2diYubioaLn4BdlKUNHTQ8gB0MTmRaLp%2BdgC%2B-nYATo4gzUgA2hnVEmFxaAC6iNgogcFQfo3%2BlBC82K2IoJCwCCNgAJ4AYmBsZfsLAOYoItibnYO6p5t5QWAhiJ1B4FB9jgDCcC-DBsjo3gQiKRpMU5AwwCAALYAIzQzXYAAYJGdtmBav4wp8XmwAIy1IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function useStateWithSideEffect<T>(initialState: T, sideEffect?: (newValue: T) => void) {
  let value = initialState;

  const setValue = (newValue: T) => {
    value = newValue;

    sideEffect?.(newValue);
  }

  return [value, setValue] as const;
}


declare function anyFn(...args: any[]): any;

const [count, setCount] = useStateWithSideEffect<number>(0, anyFn);

setCount(1);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-arguments": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Type argument not to be flagged

### Actual Result

Type argument is flagged

### Additional Info

I was updating one of the repos I'm maintaining to the latest version of ts eslint and noticed this new error.
The original function is slightly longer than this, but this is the most minimal repro I was able to create.

`count` is typed as `0` without it, so I'm widening to `number` and getting this error.

To be honest, I'm not sure of the exact reason for this (I guess it has to do with #11918), which is why the PR title is so vague (sorry for that!)

**Comment by StyleShit** [comment]
Ok, I guess it has to do with any `any` (pun intended):


```TS
function identity<T>(value: T) { 
  return value;
}

const map = identity<Map<string, number>>(new Map());

// `new Map()` results in `Map<any, any>`
```

[playground](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=GYVwdgxgLglg9mABDAJgUzLKBPAPAFQD4AKANwEMAbENALkXwEpEBvRAKEUQCc0oRuSCtTQBudgF927CAgDOURAFtyAB0QBeZOkwwcuALJrcC7jDABzADSIwIJQCM03QiTBoA7oiOrijRqJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

This could be solved ofc by moving the type parameters to the `new Map()` call

___ 

I'm also getting the same error with a generic type when the passed type argument has only optional parameters:

```TS
type Wrapper<T = Record<string, any>> = {
  args: T;
}

type Works = Wrapper<{ a: number }>;
type Fails = Wrapper<{ a?: number }>;
```

[playground](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=C4TwDgpgBA6gTgQzJOAeAKlAvFAShAYwHs4ATVAZ2DgEsA7AcwBooE6QA%2BD7KAbwCgorOAwoAuKOgDc-AL79%2BoSLBIBrCj3hIUqXqwl0ArgFsARhDhRZHGUugAxBDQA2GnFuQXdrAPwGT5pbWUkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by StyleShit** [comment]
Oops, didn't see #12158 🤦
